fixed#2762 Lift vertex count restriction for bipartite

devlohani99 opened the pull request:
<!-- Please disclose any use of LLMs (ChatGPT, Copilot, Gemini, etc.) during the preparation of this PR. -->
I used Copilot (Anthropic)  to help for preparation for this PR .

<!-- The following text must be left intact and the box checked before the PR can be merged. -->
- [x] By submitting this pull request, I assign the copyright of my contribution to _The igraph development team_.

If the number of possible edges is too big, the code now uses a special function called gnp_bipartite_large that handles large graphs without crashing. I made sure the direction of edges (like incoming, outgoing, or both) works properly in both normal and large graph cases. fixed some math calculations to use safer numbers, so there are no errors due to large values.



## Related Issues
Fixes #2762

szhorvat commented:
Thanks for the PR. Next time please do let us know _before_ you start working on an issue, so we can coordinate.

> I used Copilot (Anthropic) to help for preparation for this PR .

Did you use it to generate code, and if yes, to what extent?

> I made sure the direction of edges (like incoming, outgoing, or both) works properly in both normal and large graph cases. fixed some math calculations to use safer numbers, so there are no errors due to large values.

Can you please elaborate on what exact changes you made, and why?

I see multiple changes in the original implementation, and it is unclear to me what you are trying to do. 

---

I see that tests are currently failing. Can you investigate the reason and make sure that tests pass before marking the PR as ready for review?

devlohani99 commented:
Thank you for your feedback and for reviewing my PR.
On the use of Copilot/Anthropic:
I used Copilot (Anthropic) as an assistant for code suggestions and to help understand the project structure, but I reviewed and edited all code myself, Copilot mainly used for boilerplate and reference.

devlohani99 commented:
I've fixed the issues with the code and all CI checks are now passing successfully

szhorvat commented:
> I've fixed the issues with the code and all CI checks are now passing successfully

You removed interruption handling, which eliminates an important feature.

When proposing a change like this, please be explicit about what you are doing (don't leave it to the reviewer to notice the code change, mention it!) and explain why you want to do this. 

In this case, interruption handling must not be removed.

> > I made sure the direction of edges (like incoming, outgoing, or both) works properly in both normal and large graph cases. fixed some math calculations to use safer numbers, so there are no errors due to large values.
> 
> Can you please elaborate on what exact changes you made, and why?
> 
> I see multiple changes in the original implementation, and it is unclear to me what you are trying to do.

Can you please answer this question, and write a summary, in your own words, of what changes you made to the existing part of the code?

devlohani99 commented:
Thanks for the feedback — you're right, I should’ve explained the removal of interruption handling more clearly.

Interruption Handling:
I removed it earlier due to build issues with IGRAPH_ALLOW_INTERRUPTION_LIMITED, but I now realize it's important. I'll reintroduce it using IGRAPH_ALLOW_INTERRUPTION() inside the loop, following igraph's standard approach.

Edge Direction:
Updated edge creation to match igraph_full_bipartite:

Undirected: i → j+n1
Directed:
OUT: i → j+n1
IN: j+n1 → i
ALL: both directions
Overflow Protection: Added checks with IGRAPH_SAFE_ADD, and a termination condition for cases with very small p to avoid overflow.

I'll update the PR to restore proper interruption support while keeping the other fixes intact.

devlohani99 commented on src/misc/bipartite.c:
```diff
@@ -22,7 +22,6 @@
 */
 
 #include "igraph_bipartite.h"
-
```
alright

devlohani99 commented:
@szhorvat  The tests test::igraph_bipartite_game and test::igraph_bipartite_projection are failing with 'Subprocess aborted'. 

I've removed IGRAPH_CHECK from vector_int_push_back calls and added /* reserved */ comments as requested.

 Could you please clarify which vector operations in the function should have IGRAPH_CHECK and which shouldn't? 

This would help me fix the test failures correctly.

szhorvat commented:
I just rebased this on the latest `master`. Let me know if you need help pulling this change on your side.

I expect to have time to make progress with this PR next week. Let me know if you are still interested and available.

Until then, let's see if Copilot can help with a review.

devlohani99 commented:
> I just rebased this on the latest `master`. Let me know if you need help pulling this change on your side.
> 
> I expect to have time to make progress with this PR next week. Let me know if you are still interested and available.
> 
> Until then, let's see if Copilot can help with a review.

Yes, I’m still interested and available to continue working on this PR. I’ve updated my local branch to match the latest changes.

szhorvat commented on src/misc/bipartite.c:
```diff
@@ -482,28 +485,28 @@ igraph_error_t igraph_full_bipartite(igraph_t *graph,
 
         for (igraph_integer_t i = 0; i < n1; i++) {
             for (igraph_integer_t j = 0; j < n2; j++) {
-                VECTOR(edges)[ptr++] = i;
-                VECTOR(edges)[ptr++] = n1 + j;
+                igraph_vector_int_push_back(&edges, i); /* reserved */
+                igraph_vector_int_push_back(&edges, n1 + j); /* reserved */
```
The first step to cleaning up this PR is to focus only on the necessary functionality, and eliminate all other changes, such as this one.

This change in fact breaks functionality, as the `edges` vector is already at its final size. We don't want to push to the end.

But my point is: Please do not modify code that is unrelated to the main purpose of this PR, as it makes review difficult. Let's make this PR focused!

devlohani99 replied:
Thanks for the guidance!
//i  removed the unintended push_back calls in 
igraph_full_bipartite().

szhorvat commented:
I'll try to get to this soon, maybe today. Thanks for all your work!

How would you like to be credited? We usually use name + GitHub handle, as in "Dev Lohani @devlohani99".

szhorvat commented:
@ntamas, this is good to be merged, but my GitHub is too slow tonight to even get a usable "merge" button. Please go ahead.
